### PR TITLE
Admin: Request user confirmation before triggering reconnect by url

### PR DESCRIPTION
**Pull request by fgiannar**
This PR adds a confirmation dialog so that the user can confirm their intention before triggering the reconnect process by url.

#### Changes proposed in this Pull Request:

* Adds a new `ReconnectModal` with corresponding unit tests and usage instructions
* Handles `reconnect` routing using a more declarative approach, per React guidelines, by adding it to Routes in `admin.js`
* Displays the reconnect confirmation modal on `reconnect` route and redirects to `dashboard` upon dismissal or when reconnection is triggered in `main.js`.
* Upon reconnect modal confirmation the `confirm_reconnect_modal` click action is tracked.

#### Jetpack product discussion
p9dueE-1NT-p2

#### Testing instructions:
**Unit tests**
* Run `yarn test-client && yarn test-gui` and verify all tests pass

**Reconnect flow with confirmation modal**
* Setup Jetpack in your testing site
* Visit `admin.php?page=jetpack#/reconnect` and verify the reconnect modal shows up
* Click `Cancel` and verify the reconnect modal hides and the current route is `admin.php?page=jetpack#/dashboard`
* Visit `admin.php?page=jetpack#/reconnect` and verify the reconnect modal shows up
* Click `Reconnect Jetpack` and verify the reconnect process is triggered, the current route is `admin.php?page=jetpack#/dashboard` and the modal is hidden

#### Proposed changelog entry for your changes:
* Admin: Request user confirmation before triggering reconnect by url

### CODE REVIEW

File: _inc/client/components/reconnect-modal/test/component.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+			class ReconnectModalMock extends ReconnectModal {
+				constructor( props ) {
+					super( props );
+					this.clickReconnectSite = clickReconnectSite;
+				}
+			}
+			// We need to set the testProps again here, to make sure they are not affected by
+			// other tests running in between.
+			Object.assign( testProps, currentTestProps );
```
Comment (zinigor) [inline _inc/client/components/reconnect-modal/test/component.js:97]: Would using a `before` statement help here? If we don't just set the `testProps` once, but do it instead in a `before` function, they would be the same as the initial ones.

Comment (fgiannar) [reply _inc/client/components/reconnect-modal/test/component.js]: @zinigor I have added the `before` approach, but in order to get the intended test properties, aka inside every describe: defaults + **only** current block changes, I had to also add nested `before` blocks.
Let me know if that approach works or if there's a better alternative.
Many thanks for reviewing this PR!

Comment (zinigor) [reply _inc/client/components/reconnect-modal/test/component.js]: Thanks! This is exactly the approach that I had in mind. This way we're clearly separating test data preparation from other `describe` blocks, so we don't have any side effects from other tests.

**Review by zinigor — APPROVED** [review]
Works well, let's merge! Thanks!

**Pull request merged**